### PR TITLE
feat(tag-input): Add TagInput component

### PR DESCRIPTION
**Pull request by annelhote**

### CODE REVIEW

File: example-ts/src/pages/components/TagInput.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+          <Link>TagInput - TagInput</Link>
+        </Breadcrumb>
+      </Row>
+      <Row gutters>
+        <Col xs={12}>
+          <Title as="h1">TagInput - TagInput</Title>
+          <Text>
+            Le tag catégorise/classe/organise les contenus à l'aide de mots-clés. Il aide les utilisateurs à rechercher et à trouver facilement une information.
+          </Text>
+          <Playground code={code} scope={{ TagInput }} defaultShowCode />
```
Comment (annelhote) [inline example-ts/src/pages/components/TagInput.tsx]: Obviously I can not import "TagInput" from `@dataesr/react-dsfr` ! But how ?

**Review by folland87 — COMMENTED** [review]
Je ne sais pas trop dire si c'est le cas, mais si ce composant n'est pas accessible, faut-il vraiment le mettre ici et le disseminer dans nos apps ?

File: src/components/TagInput/TagInput.tsx [hunk 2]
```diff
@@ -17,6 +18,7 @@ export const TagInput = forwardRef<HTMLInputElement, TagInputProps>(({
   hint,
   label,
   onTagsChange,
+  placeholder = 'Press ENTER to search for serveral terms',
   tags,
 }) => {
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx]: TagInputProps merges TagInputBaseProps with InputHTMLAttributes<HTMLInputElement>
It means any attribute accepted by input tag would be accepted for the TagInput. To take advantage of it, do not use placeholder or any of InputHTMLAttributes props, instead, get remaining props with ...props and pass it directly to the input.

File: src/components/TagInput/TagInput.tsx [hunk 3]
```diff
@@ -17,6 +18,7 @@ export const TagInput = forwardRef<HTMLInputElement, TagInputProps>(({
   hint,
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx:18]: forwardRef is used but the ref is not used

File: src/components/TagInput/TagInput.tsx [hunk 4]
```diff
@@ -17,6 +18,7 @@ export const TagInput = forwardRef<HTMLInputElement, TagInputProps>(({
...
   label,
   onTagsChange,
+  placeholder = 'Press ENTER to search for serveral terms',
   tags,
 }) => {
   const [input, setInput] = useState('');
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx:25]: In handleKeyDown, do not use e.keyCode but e.key
I think for accessibility, Tab should not handle new key

File: src/components/TagInput/TagInput.tsx [hunk 5]
```diff
@@ -42,19 +44,21 @@ export const TagInput = forwardRef<HTMLInputElement, TagInputProps>(({
...
-        type="text"
-        value={input}
-        label={label}
         hint={hint}
+        label={label}
         onChange={(event: FormEvent<HTMLInputElement>) => setInput((event?.target as HTMLTextAreaElement)?.value)}
         onKeyDown={(event: KeyboardEvent<HTMLInputElement>) => handleKeyDown(event)}
+        placeholder={placeholder}
+        type="text"
+        value={input}
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx:61]: ref={ref}
{...props}

see previous comment

File: src/components/TagInput/TagInput.tsx [hunk 6]
```diff
@@ -42,19 +44,21 @@ export const TagInput = forwardRef<HTMLInputElement, TagInputProps>(({
...
       />
       <TagGroup>
         {(values ?? []).map((tag) => (
           <Tag
-            key={tag}
             className="fr-mr-1w"
+            key={tag}
             onClick={() => handleDeleteClick(tag)}
+            title={`Tag ${tag}`}
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx]: I think this is useless for accessibility but not sure, havn't seen that in other tagInput components

**Review by folland87 — CHANGES_REQUESTED** [review]
J'ai pas vu grand chose de mieux niveau accessibilité chez les autres ! :D
Pour une version non controllée, il faudrait ajouter un input type="hidden"
Sinon coté ui, j'ai vu plus des trucs dans le genre de l'image ci jointe avec les tags a l'intérieur.
Capture d’écran du 2024-03-20 18-23-44
En vrai ca pourrait même être parametrable !

File: src/components/TagInput/TagInput.tsx (hunk 6 above)
Comment (annelhote) [reply src/components/TagInput/TagInput.tsx]: It seems to be necessary for accessibility but the text should be something like "Word (tag) selected, press Enter to delete it"

File: src/components/TagInput/TagInput.tsx (hunk 5 above)
Comment (annelhote) [reply src/components/TagInput/TagInput.tsx:61]: I simply removed forwardRef

**Comment by annelhote** [comment]
1. As we are re-coding this component from projects to projects, yes it seems to be a good place to share it here and have a unique place to fix / improve it.
2. As it is officially an extended version of the DSFR, I think that the not so perfect accessibility of a component should not be a brake on creating new component.

File: src/components/TagInput/TagInput.tsx [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        onChange={(event: FormEvent<HTMLInputElement>) => setInput((event?.target as HTMLTextAreaElement)?.value)}
+        onKeyDown={(event: KeyboardEvent<HTMLInputElement>) => handleKeyDown(event)}
+        type="text"
+        value={input}
+        {...props}
+      />
+      <TagGroup>
+        {(values ?? []).map((tag) => (
+          <Tag
+            className="fr-mr-1w"
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx:68]: as="button"

File: src/components/TagInput/TagInput.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            key={tag}
+            onClick={() => handleDeleteClick(tag)}
+            title={`Mot (${tag}) sélectionné, cliquer sur la touche "Entrée" ou la touche "Espace" pour le supprimer.`}
+          >
+            {tag}
+            <span className='fr-icon-close-line' />
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx]: better not to use a span here 
best alternative : add css ```.custom-dismissible-tag {
  &::after {
    flex: 0 0 auto;
    display: inline-block;
    vertical-align: calc((.75em - var(--icon-size)) * .5);
    background-color: currentColor;
    width: var(--icon-size);
    height: var(--icon-size);
    -webkit-mask-size: 100% 100%;
    mask-size: 100% 100%;
    -webkit-mask-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAyNCAyNCIgd2lkdGg9IjI0IiBoZWlnaHQ9IjI0Ij48cGF0aCBkPSJtMTIgMTAuNiA0Ljk1LTQuOTYgMS40IDEuNEwxMy40MiAxMmw0Ljk2IDQuOTUtMS40IDEuNEwxMiAxMy40MmwtNC45NSA0Ljk2LTEuNC0xLjRMMTAuNTggMTIgNS42MyA3LjA1bDEuNC0xLjR6Ii8+PC9zdmc+);
    mask-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAyNCAyNCIgd2lkdGg9IjI0IiBoZWlnaHQ9IjI0Ij48cGF0aCBkPSJtMTIgMTAuNiA0Ljk1LTQuOTYgMS40IDEuNEwxMy40MiAxMmw0Ljk2IDQuOTUtMS40IDEuNEwxMiAxMy40MmwtNC45NSA0Ljk2LTEuNC0xLjRMMTAuNTggMTIgNS42MyA3LjA1bDEuNC0xLjR6Ii8+PC9zdmc+);
    --icon-size: 1rem;
    content: "";
    margin-left: .25rem;
    margin-right: -.25rem;
  }
}```  in DismissableTag and use DismissableTag here. It is a button by default
Ok alternative : use the css directly in InputTag and set as="button"
Not so good alternative : add fr-icon--sm and aria-hidden

File: src/components/TagInput/TagInput.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+  const handleDeleteClick = (tag: string) => {
+    const newValues = [...(values ?? []).filter((el) => el !== tag)];
+    setValues(newValues);
+    onTagsChange(newValues);
+  };
+
+  return (
+    <div>
+      <TextInput
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx]: Not sure it is a good idea to use a TextInput here espacially if features of it like message, messagetype are not used.
At least disableAutoValidation should be set to true

File: src/components/TagInput/TagInput.tsx [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        hint={hint}
+        label={label}
+        onChange={(event: FormEvent<HTMLInputElement>) => setInput((event?.target as HTMLTextAreaElement)?.value)}
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx:57]: onChange={(event: FormEvent<HTMLInputElement>) => setInput((event?.target as HTMLInputElement)?.value)}

File: src/components/TagInput/TagInput.tsx [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        value={input}
+        {...props}
+      />
+      <TagGroup>
+        {(values ?? []).map((tag) => (
+          <Tag
+            className="fr-mr-1w"
+            key={tag}
+            onClick={() => handleDeleteClick(tag)}
+            title={`Mot (${tag}) sélectionné, cliquer sur la touche "Entrée" ou la touche "Espace" pour le supprimer.`}
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx]: aria-label instead of title

Comment (folland87) [reply src/components/TagInput/TagInput.tsx]: Possibly a props to set it in other languages ?

**Review by folland87 — CHANGES_REQUESTED** [review]

File: src/components/TagInput/TagInput.tsx (hunk 11 above)
Comment (annelhote) [reply src/components/TagInput/TagInput.tsx]: https://dataesr.github.io/react-dsfr/?path=/docs/composants-tag-tag--alt

File: src/components/Tag/Tag.tsx [hunk 12]
```diff
@@ -89,13 +85,13 @@ export const DissmissibleTag = polyRef<"button", TagProps, OnlyAs<"button" | "a"
   size,
   ...props
 }, ref) => {
-  const _classes = cn('fr-tag--dismiss', getTagClasses({ className, color, icon, iconPosition, size }));
+  const _classes = cn('fr-tag--dismiss', 'custom-dismissible-tag', getTagClasses({ className, color, icon, iconPosition, size }));
```
Comment (folland87) [inline src/components/Tag/Tag.tsx]: normalement vient remplacer fr-tag--dismiss

File: src/components/Tag/Tag.tsx [hunk 13]
```diff
@@ -17,21 +19,15 @@ const getTagClasses = ({ className, color, icon, iconPosition, size }: TagProps)
...
 
-export type SelectableTagProps = {
-  className?: Argument;
-  color?: ColorFamily;
-  icon?: string;
-  iconPosition?: "left" | "right";
-  selected: boolean;
-  size?: "md" | "sm";
-}
+export type SelectableTagProps = Merge<TagProps, { selected: boolean; }>;
```
Comment (folland87) [inline src/components/Tag/Tag.tsx]: Voir la fonction Merge => TagProps & { selected: boolean } suffirait ici

File: src/components/Tag/styles.scss [hunk 14]
```diff
@@ -29,7 +29,7 @@ $color-families: green-tilleul-verveine,
 }
 
 
-// Dissmissable tags
+// Dismissable tags
 @each $color in $color-families {
   .fr-tag.fr-tag--#{$color}.fr-tag--dismiss {
```
Comment (folland87) [inline src/components/Tag/styles.scss]: .fr-tag.fr-tag--#{color}.custom-dismissable-tag

**Review by folland87 — CHANGES_REQUESTED** [review]

File: src/components/Tag/Tag.tsx (hunk 12 above)
Comment (annelhote) [reply src/components/Tag/Tag.tsx]: Thanks, I had a doubt writing this

File: src/components/TagInput/TagInput.tsx [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+  };
+
+  return (
+    <div className="fr-input-group">
+      <label className={cn('fr-label')} htmlFor={id}>
+        {label}
+        {hint && <span className={cn('fr-hint-text')}>{hint}</span>}
+      </label>
+      <div
+        className={cn('fr-input', styles['taginput-wrapper'])}
```
Comment (folland87) [inline src/components/TagInput/TagInput.tsx]: not sure if taginput-wrapper and taginput-input are necessary, at least not all nested classes to handle focus

**Review by folland87 — APPROVED** [review]

**Pull request merged**